src/WebJobs.Script.WebHost/Security/Authentication/AuthenticationLevelHandler.cs

```diff
@@ -39,10 +40,43 @@ internal class AuthenticationLevelHandler : AuthenticationHandler<Authentication
 
         protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
         {
-            // Get the authorization level for the current request
+            var requestIdentities = await GetRequestIdentities();
+            if (requestIdentities.Any())
+            {
+                return AuthenticateResult.Success(new AuthenticationTicket(new ClaimsPrincipal(requestIdentities), Scheme.Name));
+            }
+            else
+            {
+                return AuthenticateResult.NoResult();
+            }
+        }
+
+        private async Task<IEnumerable<ClaimsIdentity>> GetRequestIdentities()
```

Comment:
>IEnumerable<ClaimsIdentity> [](start = 27, length = 27)

`IEnumerable<T>` is one of those types you have to be careful about using. Instead of returning an `IEnumerable<ClaimsIdentity>`, you should return an `ICollection<ClaimsIdentity>`. That way the caller can do `requestIdentities.Count` instead of `requestIdentities.Any()` which has more perf overhead. It also ensures this enumerator doesn't get run multiple times.

Reply:
As of right now, this method is only being used once, and it is by the above method, which only uses it for a `.Any()` call and to feed it into `ClaimsPrincipal`, which takes an `IEnumerable<ClaimsIdentity>`. I actually would suspect that calling `.Any()` once and then being passed directly to the `ClaimsPrincipal` might be more performant than converting it to a list just to do the count operation.

However, since this method may be used in other ways in the future, it might still be worthwhile to still convert this to a list.

---
In reply to: [181941843](https://github.com/Azure/azure-functions-host/pull/2684#discussion_r181941843) [](ancestors = 181941843)